pkg/storage/mysql/mysqlstore.go

```diff
@@ -0,0 +1,268 @@
+	}
+
+	store := &sqlDBStore{
+		db:        newDBConn,
+		tableName: tableName}
+
+	p.dbs[name] = store
+
+	return store, nil
+}
+
+// Close closes the provider.
+func (p *Provider) Close() error {
+	p.Lock()
+	defer p.Unlock()
+
+	for _, store := range p.dbs {
+		err := store.db.Close()
+		if err != nil {
+			return fmt.Errorf(failToCloseProviderErrMsg+": %w", err)
+		}
+	}
+
+	if err := p.db.Close(); err != nil {
+		return err
+	}
+
+	p.dbs = make(map[string]*sqlDBStore)
+
+	return nil
+}
+
+// CloseStore closes a previously opened store
+func (p *Provider) CloseStore(name string) error {
+	p.Lock()
+	defer p.Unlock()
+
+	if p.dbPrefix != "" {
+		name = p.dbPrefix + "_" + name
+	}
+
+	store, exists := p.dbs[name]
+	if !exists {
+		return storage.ErrStoreNotFound
+	}
+
+	delete(p.dbs, name)
+
+	return store.db.Close()
+}
+
+// Put stores the key and the value
+func (s *sqlDBStore) Put(k string, v []byte) error {
+	if k == "" {
+		return errors.New("key is mandatory")
+	}
+
+	//nolint: gosec
+	// create upsert query to insert the record, checking whether the key is already mapped to a value in the store.
+	createStmt := "INSERT INTO " + s.tableName + " VALUES (?, ?) ON DUPLICATE KEY UPDATE value=?"
+	// executing the prepared insert statement
+	_, err := s.db.Exec(createStmt, k, v, v)
+	if err != nil {
+		return fmt.Errorf("failed to insert key and value record into %s %w ", s.tableName, err)
+	}
+
+	return nil
+}
+
+// Get fetches the value based on key
+func (s *sqlDBStore) Get(k string) ([]byte, error) {
+	if k == "" {
+		return nil, errors.New("key is mandatory")
+	}
+
+	var value []byte
+	//nolint: gosec
+	// select query to fetch the value by key
+	err := s.db.QueryRow("SELECT `value` FROM "+s.tableName+" "+
+		" WHERE `key` = ?", k).Scan(&value)
+	if err != nil {
+		if strings.Contains(err.Error(), sqlDBNotFound) {
+			return nil, storage.ErrValueNotFound
+		}
+
+		return nil, fmt.Errorf("failed to get row %w", err)
+	}
+
+	return value, nil
+}
+
+type sqlDBResultsIterator struct {
+	resultRows *sql.Rows
+	err        error
+}
+
+func (s *sqlDBStore) CreateIndex(createIndexRequest storage.CreateIndexRequest) error {
+	// TODO: Issue-36 Implement Create Index
+	return nil
+}
+
+func (s *sqlDBStore) Query(findQuery string) (storage.ResultsIterator, error) {
+	resultRows, err := s.db.Query(findQuery)
+	if err != nil && resultRows == nil {
+		return nil, fmt.Errorf("failed to query rows %w", err)
+	} else if err := resultRows.Err(); err != nil {
+		return nil, fmt.Errorf("failed to get resulted rows %w", err)
+	}
+
+	return &sqlDBResultsIterator{resultRows: resultRows}, nil
+}
+
+// Key returns the key of the current key-value pair.
+func (i *sqlDBResultsIterator) Key() (string, error) {
+	var res result
+
+	err := i.resultRows.Scan(&res.key, &res.value)
+	if err != nil {
+		return "", fmt.Errorf("failed to scan the SQL rows while getting key: %w", err)
+	}
+
+	return res.key, nil
```

Comment:
With this implementation, two successive calls to `Key()` and `Value()` in any order will not correspond to the same db row.

Looks like `sqlDBResultsIterator` needs to hold on to the `result` until `Next()` is called

Reply:
updated